### PR TITLE
Blocks: Propose a server side rendering implementation

### PR DESCRIPTION
**Pull request by youknowriad**
This is also called dynamic blocks and is useful for blocks like embed, comments blocks, ...
This is still in progress but feedback is appreciated.

**Testing instructions**

 - Insert this `<!-- wp:core/html -->anything<!-- /wp:core/html -->` to a post content (old editor) and preview this post, you should see your block replaced by `THIS IS AWESOME`

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
...
+
+/**
+ * Registers a block.
+ *
+ * @param  string   $name   Block name including namespace.
+ * @param  function $render Render function called to generate the HTML output
+ */
+function register_block( $name, $render ) {
+	global $registered_blocks;
+	$registered_blocks[ $name ] = $render;
```
Comment (aduth) [inline index.php]: I don't want to preemptively assume requirements, but do you think we'd need a block to register options aside from just a render?

File: index.php [hunk 2]
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
...
+/**
+ * Renders the dynamic blocks into the post content
+ *
+ * @since 0.1.0
+ *
+ * @param  string $content Post content
+
+ * @return string          Updated post content
+ */
+function do_dynamic_blocks( $content ) {
```
Comment (aduth) [inline index.php]: This function could do for some comments; various indices are hard to follow.

File: index.php [hunk 3]
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
...
+	global $registered_blocks;
+	$open_matcher = '/<!--\s*wp:([a-z](?:[a-z0-9\/]+)*)\s+((?:(?!-->).)*)-->.*<!--\s*\/wp:\g1\s+-->/';
+	preg_match_all( $open_matcher, $content, $matches, PREG_OFFSET_CAPTURE );
+	$replacements = array();
+	$offset = 0;
+	$new_content = $content;
+	foreach ( $matches[ 0 ] as $index => $block_match ) {
+		$block_name = $matches[ 1 ][ $index ][ 0 ];
+		if ( isset( $registered_blocks[ $block_name ] ) ) {
```
Comment (aduth) [inline index.php]: Maybe worth an early continue to avoid indentation:

```php
if ( ! isset( $registered_blocks[ $block_name ] ) ) {
	continue;
}
```

File: index.php [hunk 4]
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
...
+			$block_length = mb_strlen( $block_match[ 0 ] );
```
Comment (aduth) [inline index.php]: I don't know the answer, but is `mb_strlen` the safe default over `strlen` for multi-byte content?

File: index.php [hunk 5]
```diff
@@ -247,3 +316,7 @@ function the_gutenberg_project() {
 	</div>
 	<?php
 }
+
+register_block( 'core/html', function( $attributes ) {
```
Comment (aduth) [inline index.php]: Seems temporary, but just for clarity, the anonymous function callable is only available in PHP versions 5.3.0 and newer ([reference](http://php.net/manual/en/functions.anonymous.php)). WordPress supports as far back as 5.2.4.

File: index.php [hunk 6]
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
...
+ *
+ * @since 0.1.0
+ *
+ * @param  string $content Post content
+
+ * @return string          Updated post content
+ */
+function do_dynamic_blocks( $content ) {
+	global $registered_blocks;
+	$open_matcher = '/<!--\s*wp:([a-z](?:[a-z0-9\/]+)*)\s+((?:(?!-->).)*)-->.*<!--\s*\/wp:\g1\s+-->/';
```
Comment (aduth) [inline index.php]: Would we want the registered block to be able to make use of the content between the comment delimiters?

File: index.php (hunk 1 above)
Comment (youknowriad) [inline index.php]: I tried to think about this, but I didn't find any use-case for now. Though, we don't know the future maybe options is preferable.

File: index.php (hunk 6 above)
Comment (youknowriad) [inline index.php]: I know there were some discussions about this, but IMO for consistency, Dynamic blocks should not contain any data between the delimiters. Their `save` function should return `null`.

File: index.php (hunk 4 above)
Comment (youknowriad) [inline index.php]: In my PHP souvenirs, it is, but I'm not in my comfort zone anymore in PHP. Other 👀 here could be useful.

File: index.php (hunk 5 above)
Comment (youknowriad) [inline index.php]: Oh right 👍

**Comment by mtias** [comment]
Thanks for starting this.

File: index.php (hunk 2 above)
Comment (mtias) [inline index.php]: Is the specificity of "dynamic" something we want to expose in the function name? I'd expect we'll add hooks here for interacting with individual blocks (say I want to filter all quote blocks for some reason). Performance here will be very important as well.

**Comment by youknowriad** [comment]
I'm thinking to leverage this for the `embed` block. But to do so, we need to change the way we store attributes (use HTML like syntax). 

What do you think if 
 - we remove the dummy "register_block" call. 
 - Try to merge this proposal
 - do a follow-up PR to change the attributes' syntax 
 - and then another PR to add the server side embed block. 

(I like splitting/Iterating :) )

**Comment by aduth** [comment]
I'm not totally clear what you mean by changing attribute syntax, but I'd be fine with merging this and iterating on it in subsequent pull requests since it has no impact on existing behavior for the moment.

**Comment by nylen** [comment]
This logic looks fairly complicated and needs unit tests as soon as possible.  See https://pippinsplugins.com/unit-tests-wordpress-plugins-setting-up-testing-suite/ for example.

I think we should do more to unify our block registration APIs between the client and server (require the same rules for block slugs, for example).  I also wonder if we need or want some way to guarantee that if a block is registered on the server, then it's also registered on the client (and the relevant details match up).

I'm not sure those need to be done in this PR, but I'm also not sure why we need to merge this immediately.  Are we ready to use it yet?

**Comment by BE-Webdesign** [comment]
What does the server side API need to do and why does it need to do it, would be good questions to answer before we get too attached to this. I am fine with merging, but I think the original JS API could use improvements right now as well, and we should circle back to refining the API at some point.

**Comment by westonruter** [comment]
Additionally, I think there is also the question as to whether the server-side API should also be able to facilitate rendering of blocks for displaying inside of the visual editor, similar to what the Shortcake plugin does. A good case for server-side rendering for the block in the editor is both for legacy content (e.g. shortcodes) but also for displaying dynamic blocks that are not feasible to provide pure JS renderings without duplicating a lot of logic.

**Comment by youknowriad** [comment]
@aduth 

> I'm not totally clear what you mean by changing attribute syntax

moving from `attr:value` to `attr="value"` or `{ attr: "value" }`. We can't store URLs for example for now.

@nylen 

> This logic looks fairly complicated

I don't think so

> needs unit tests as soon as possible

I agree

> I think we should do more to unify our block registration APIs between the client and server

Sounds good for the validation

> . I also wonder if we need or want some way to guarantee that if a block is registered on the server, then it's also registered on the client

I don't think so, we only need the dynamic blocks on the server, why forcing plugin authors to add something optional

>  but I'm also not sure why we need to merge this immediately

I'm not so impatient to merge this, I was waiting for review and you raise some good points here. But  I like to split the work to smaller PRs.

> Are we ready to use it yet?

Yes we are, we can use this to generate the HTML for the embed block for example.


@BE-Webdesign 

> What does the server side API need to do, why does it need to do it

The only thing that it needs to do for now is rendering dynamic blocks and we currently have a dynamic block on the client side (The embed block). I'm trying to start the server side implementation to tackle this particular task. We'll iterate and see if we need any other server side API.

@westonruter 

> Additionally, I think there is also the question as to whether the server-side API should also be able to facilitate rendering of blocks for displaying inside of the visual editor

I'm not sure I understand! Are you saying we should generate a form in the visual editor server-side? I would personally strongly argue against that and instead, build a specific "nice" UI for each block. But I understand this could be helpful for mobile apps maybe. Anyway, this is a separate task and for now, I'm trying to address the server-side rendering of dynamic blocks.

File: index.php (hunk 6 above)
Comment (mtias) [inline index.php]: We shouldn't prescribe what save returns, because it may make sense to return a link, or a short message. Consider a block that renders the last 5 recent posts from a certain category. It may be good that when the dynamic markup can't be generated, it returns a link to `/blog/category/:term` with some text saying "View latest posts on :term".

**Comment by nylen** [comment]
> I'm not so impatient to merge this, I was waiting for review and you raise some good points here. But I like to split the work to smaller PRs.

I agree, I guess it makes me a bit nervous that we are building a copy of an API that already exists on the client side.  We need to be careful that it works the same way as much as possible.  To me that means we should follow a similar process here to #289, including the same test cases.  So I'd prefer that we finish and merge #617 (PHPUnit) first, then add test cases here.

**Comment by youknowriad** [comment]
Server side tests in place, any other concern here?

**Comment by nylen** [comment]
This was also present on #617, but I'm seeing a lot of warnings in the build log:

> PHP Warning:  filemtime(): stat failed for /home/travis/build/WordPress/gutenberg/i18n/build/index.js in /home/travis/build/WordPress/gutenberg/index.php on line 161

Example: https://travis-ci.org/WordPress/gutenberg/jobs/228752501

Also, from https://github.com/WordPress/gutenberg/pull/617#issuecomment-299185065, was this ever addressed?

> Take a close look at each of the jobs, there are two showing "green" that they passed, this is false, there's lots of errors on those jobs.

Let's fix the warnings here and make sure that when the build fails it actually fails like it should.  I'll do a separate PR against this branch to test out the second thing.

**Comment by youknowriad** [comment]
Oh! I haven't seen those errors. I thought he was talking about the failing jobs. At that moment three of five jobs were failing.

File: phpunit/class-dynamic-blocks-render-test.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	 * @return string             Block output.
+	 */
+	function render_dummy_block( $attributes ) {
+		return $attributes['value'];
+	}
+
+	/**
+	 * Successfull dynamic block rendering
+	 */
+	function test_register_block() {
```
Comment (nylen) [inline phpunit/class-dynamic-blocks-render-test.php]: In this PR there have been issues with matching the content inside the block delimiters. Let's get another test similar to this one but testing out content inside blocks as well:

```html
before
<!-- wp:core/dummy value:a1 -->this should be ignored<!-- /wp:core/dummy -->
between
<!-- wp:core/dummy value:a2 -->this should be ignored also<!-- /wp:core/dummy -->
after
```

File: index.php [hunk 8]
```diff
@@ -27,6 +27,121 @@ function gutenberg_menu() {
...
+	}
+
+	$slug_matcher = '/^[a-z0-9-]+\/[a-z0-9-]+$/';
+	if ( ! preg_match( $slug_matcher, $slug ) ) {
+		$message = __( 'Block slugs must contain a namespace prefix. Example: my-plugin/my-custom-block' );
+		_doing_it_wrong( __FUNCTION__, $message, '0.1.0' );
+		return;
+	}
+
+	if ( isset( $registered_blocks[ $slug ] ) ) {
```
Comment (nylen) [inline index.php]: Why not trigger a `_doing_it_wrong` here too, like we do on the client side?

File: index.php [hunk 9]
```diff
@@ -27,6 +27,121 @@ function gutenberg_menu() {
...
+/**
+ * Unregisters a block.
+ *
+ * @param  string $slug Block slug.
+ * @return array        The previous block value, if it has been
+ *                        successfully unregistered; otherwise `null`.
+ */
+function unregister_block( $slug ) {
+	global $registered_blocks;
+	if ( ! isset( $registered_blocks[ $slug ] ) ) {
```
Comment (nylen) [inline index.php]: Why not trigger a `_doing_it_wrong` here if this is set, like we do on the client side?

**Comment by nylen** [comment]
Here are the block registration test cases on the client side:

```
  blocks
    registerBlock()
      ✓ should reject numbers
      ✓ should reject blocks without a namespace
      ✓ should reject blocks with invalid characters
      ✓ should accept valid block names
      ✓ should prohibit registering the same block twice
      ✓ should store a copy of block settings
    unregisterBlock()
      ✓ should fail if a block is not registered
      ✓ should unregister existing blocks
```

Let's get all of the same tests running here too.

**Comment by youknowriad** [comment]
I've added more unit tests fc7fabe (closer to what we have on the frontend). Can I have another look @nylen?

**Comment by nylen** [comment]
There were still a couple of tests missing:

- should prohibit registering the same block twice
- https://github.com/WordPress/gutenberg/pull/586#discussion_r114985211
- verify that the functions actually make the expected changes to the blocks global variable

Other changes:

- As of 47e81bd every test method no longer needs to have an annotation.
- Changed `$registered_blocks` to `$wp_registered_blocks`; in WP, [prefix everything](https://nacin.com/2010/05/11/in-wordpress-prefix-everything/).
- Explicit `return false;` for error conditions; also test accordingly.
- You can now run `phpunit` from within a WP dev install without additional configuration or having to specify `WP_TESTS_DIR`.

**Review by nylen — APPROVED** [review]
If latest changes look OK: :shipit:
